### PR TITLE
Add handling of lat, lon coordinate queries

### PR DESCRIPTION
**Pull request by alexandervlpl**
Quick, *pure js* implementation to handle coordinate queries (#147). If a user enters valid coordinates, they get a single result with that point (as in Google Maps and other services). It's a useful feature for power users.
image

Any other queries are sent to the Provider. Happy to see this in TypeScript and otherwise improved, it's probably not ready for merging.

### CODE REVIEW

**Comment by smeijer** [comment]
Sorry for missing this. This looks great! Thanks.

File: src/coords.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
+// @ts-nocheck
+
+export function validateCoords(query) {
+    const q = query.query.trim();
```
Comment (smeijer) [inline src/coords.ts:4]:
```suggestion
    const q = query.trim();
```

File: src/SearchControl.ts [hunk 2]
```diff
@@ -373,7 +374,13 @@ const Control: SearchControl = {
     const { provider } = this.options;
 
     if (query.length) {
-      let results = await provider!.search({ query });
+      var results = [];
+      const coords = validateCoords({query});
```
Comment (smeijer) [inline src/SearchControl.ts:378]:
```suggestion
      const coords = validateCoords(query);
```

File: src/SearchControl.ts [hunk 3]
```diff
@@ -373,7 +374,13 @@ const Control: SearchControl = {
     const { provider } = this.options;
 
     if (query.length) {
-      let results = await provider!.search({ query });
+      var results = [];
```
Comment (smeijer) [inline src/SearchControl.ts:377]:
```suggestion
      const results = [];
```

File: src/SearchControl.ts [hunk 4]
```diff
@@ -373,7 +374,13 @@ const Control: SearchControl = {
...
+      const coords = validateCoords(query);
+      if (coords) {
+        results = coords;
+      } else {
+        results = await provider!.search({query});
```
Comment (smeijer) [inline src/SearchControl.ts:382]:
```suggestion
        results = await provider!.search({ query });
```

File: src/SearchControl.ts [hunk 5]
```diff
@@ -385,7 +392,13 @@ const Control: SearchControl = {
     this.resultList.clear();
     const { provider } = this.options;
 
-    const results = await provider!.search(query);
+    var results = [];
```
Comment (smeijer) [inline src/SearchControl.ts:395]:
```suggestion
    const results = [];
```

File: src/SearchControl.ts [hunk 6]
```diff
@@ -385,7 +392,13 @@ const Control: SearchControl = {
...
+    const coords = validateCoords(query);
+    if (coords) {
+        results = coords;
```
Comment (smeijer) [inline src/SearchControl.ts:398]:
```suggestion
      results = coords;
```

File: src/SearchControl.ts [hunk 7]
```diff
@@ -385,7 +392,13 @@ const Control: SearchControl = {
...
+    } else {
+        results = await provider!.search(query);
```
Comment (smeijer) [inline src/SearchControl.ts:400]:
```suggestion
      results = await provider!.search(query);
```

File: src/coords.ts [hunk 8]
```diff
@@ -0,0 +1,21 @@
+// @ts-nocheck
+
+export function validateCoords(query) {
+    const q = query.trim();
```
Comment (smeijer) [inline src/coords.ts:4]:
```suggestion
  const q = query.trim();
```

File: src/coords.ts [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+    const regex = /^(-?[0-9]*\.?\s*[0-9]*)\s*,?\s*(-?[0-9]*\.?[0-9]*)$/g;
+    var match = regex.exec(q);
```
Comment (smeijer) [inline src/coords.ts:6]:
```suggestion
    const match = regex.exec(q);
```

File: src/SearchControl.ts [hunk 10]
```diff
@@ -385,7 +392,13 @@ const Control: SearchControl = {
     this.resultList.clear();
     const { provider } = this.options;
 
-    const results = await provider!.search(query);
+    const results = [];
```
Comment (smeijer) [inline src/SearchControl.ts:395]:
```suggestion
    let results = [];
```

File: src/SearchControl.ts [hunk 11]
```diff
@@ -373,7 +374,13 @@ const Control: SearchControl = {
     const { provider } = this.options;
 
     if (query.length) {
-      let results = await provider!.search({ query });
+      const results = [];
```
Comment (smeijer) [inline src/SearchControl.ts:377]:
```suggestion
      let results = [];
```

File: src/coords.ts [hunk 12]
```diff
@@ -0,0 +1,21 @@
+// @ts-nocheck
+
+export function validateCoords(query) {
+  const q = query.trim();
+    const regex = /^(-?[0-9]*\.?\s*[0-9]*)\s*,?\s*(-?[0-9]*\.?[0-9]*)$/g;
```
Comment (smeijer) [inline src/coords.ts:5]:
```suggestion
  const regex = /^(-?[0-9]*\.?\s*[0-9]*)\s*,?\s*(-?[0-9]*\.?[0-9]*)$/g;
```

Comment (smeijer) [inline src/coords.ts:5]:
```suggestion
  const regex = /^(-?[0-9]*\.?\s*[0-9]*)\s*,?\s*(-?[0-9]*\.?[0-9]*)$/g;
```

File: src/coords.ts [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+    const match = regex.exec(q);
```
Comment (smeijer) [inline src/coords.ts:6]:
```suggestion
  const match = regex.exec(q);
```

File: src/coords.ts [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+    if (match) {
```
Comment (smeijer) [inline src/coords.ts:7]:
```suggestion
  if (match) {
```

File: src/coords.ts [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+        var lat = Number(match[1]);
```
Comment (smeijer) [inline src/coords.ts:8]:
```suggestion
    let lat = Number(match[1]);
```

File: src/coords.ts [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+        var lng = Number(match[2]);
```
Comment (smeijer) [inline src/coords.ts:9]:
```suggestion
    let lng = Number(match[2]);
```

File: src/coords.ts [hunk 17]
```diff
@@ -0,0 +1,21 @@
+// @ts-nocheck
+
+export function validateCoords(query) {
+  const q = query.trim();
+  const regex = /^(-?[0-9]*\.?\s*[0-9]*)\s*,?\s*(-?[0-9]*\.?[0-9]*)$/g;
+  const match = regex.exec(q);
+  if (match) {
+    let lat = Number(match[1]);
```
Comment (smeijer) [inline src/coords.ts:8]:
```suggestion
    const lat = Number(match[1]);
```

File: src/coords.ts [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+    let lng = Number(match[2]);
```
Comment (smeijer) [inline src/coords.ts:9]:
```suggestion
    const lng = Number(match[2]);
```

File: src/coords.ts [hunk 19]
```diff
@@ -0,0 +1,21 @@
...
+        if (( -90 < lat < 90) && ( -180 < lng < 180)) {
```
Comment (smeijer) [inline src/coords.ts:10]:
```suggestion
    if (-90 < lat < 90 && -180 < lng < 180) {
```

File: src/coords.ts [hunk 20]
```diff
@@ -0,0 +1,21 @@
...
+
+export function validateCoords(query) {
+  const q = query.trim();
+  const regex = /^(-?[0-9]*\.?\s*[0-9]*)\s*,?\s*(-?[0-9]*\.?[0-9]*)$/g;
+  const match = regex.exec(q);
+  if (match) {
+    const lat = Number(match[1]);
+    const lng = Number(match[2]);
+    if (-90 < lat < 90 && -180 < lng < 180) {
+            return [{
```
Comment (smeijer) [inline src/coords.ts:11]:
```suggestion
      return [
        {
```

File: src/coords.ts [hunk 21]
```diff
@@ -0,0 +1,21 @@
...
+                x: lng,
```
Comment (smeijer) [inline src/coords.ts:12]:
```suggestion
          x: lng,
```

File: src/coords.ts [hunk 22]
```diff
@@ -0,0 +1,21 @@
...
+                y: lat,
```
Comment (smeijer) [inline src/coords.ts:13]:
```suggestion
          y: lat,
```

File: src/coords.ts [hunk 23]
```diff
@@ -0,0 +1,21 @@
...
+                label: q,
```
Comment (smeijer) [inline src/coords.ts:14]:
```suggestion
          label: q,
```

File: src/coords.ts [hunk 24]
```diff
@@ -0,0 +1,21 @@
...
+                bounds: null,
```
Comment (smeijer) [inline src/coords.ts:15]:
```suggestion
          bounds: null,
```

File: src/coords.ts [hunk 25]
```diff
@@ -0,0 +1,21 @@
...
+                raw: {}
```
Comment (smeijer) [inline src/coords.ts:16]:
```suggestion
          raw: {}
```

File: src/coords.ts [hunk 26]
```diff
@@ -0,0 +1,21 @@
...
+            }];
```
Comment (smeijer) [inline src/coords.ts:17]:
```suggestion
      },
    ];
```

File: src/coords.ts [hunk 27]
```diff
@@ -0,0 +1,21 @@
...
+        }
```
Comment (smeijer) [inline src/coords.ts:18]:
```suggestion
    }
```

File: src/coords.ts [hunk 28]
```diff
@@ -0,0 +1,21 @@
...
+    }
```
Comment (smeijer) [inline src/coords.ts:19]:
```suggestion
  }
```

File: src/coords.ts [hunk 29]
```diff
@@ -0,0 +1,21 @@
...
+    return false;
```
Comment (smeijer) [inline src/coords.ts:20]:
```suggestion
  return false;
```

File: src/coords.ts [hunk 30]
```diff
@@ -0,0 +1,23 @@
...
+    if (-90 < lat < 90 && -180 < lng < 180) {
+      return [
+        {
+          x: lng,
+          y: lat,
+          label: q,
+          bounds: null,
+          raw: {}
+      },
+    ];
```
Comment (smeijer) [inline src/coords.ts:19]:
```suggestion
      ];
```

File: src/coords.ts [hunk 31]
```diff
@@ -0,0 +1,23 @@
...
+    const lat = Number(match[1]);
+    const lng = Number(match[2]);
+    if (-90 < lat < 90 && -180 < lng < 180) {
+      return [
+        {
+          x: lng,
+          y: lat,
+          label: q,
+          bounds: null,
+          raw: {}
```
Comment (smeijer) [inline src/coords.ts:17]:
```suggestion
          raw: {},
```

File: src/coords.ts [hunk 32]
```diff
@@ -0,0 +1,23 @@
...
+      },
```
Comment (smeijer) [inline src/coords.ts:18]:
```suggestion
        },
```

**Comment by smeijer** [comment]
Lol, I did it. Fixed it from my phone. 😆

Thanks again!

**Comment by alexandervlpl** [comment]
Awesome, thanks for improving this. :+1:

**Comment by pmaga** [comment]
This broke a standard submit: `handleSubmit: (result) => this.onSubmit(result),`. It tries to `trim` a result object that is not a string.

**Comment by alexandervlpl** [comment]
Apologies for what was clearly a bad PR, I did say it isn't ready for merging and was planning to work on it after feedback. Probably should not have submitted something so half baked. I see the concerns from @gjvoosten in #384 and they're valid.

That said, I don't think this should be abandoned because "some search providers already support coordinate searches". I think most do not, and offloading this to providers when it can be done instantly in the front end seems very inefficient.

Is it worth revisiting this, getting it TypeScripted and properly tested?

**Comment by smeijer** [comment]
No worries, I was the one that merged the PR, so I took the responsibility :) 

Happy to merge (after proper testing 😅) any other contribution you make. Thanks!

**Pull request merged**